### PR TITLE
Implementation of Qwen3-VL

### LINKED ISSUE #1063: Qwen3-VL
**Is your feature request related to a problem? Please describe.**

Currently, Tunix supports text-only Qwen3, but not multimodal Qwen3-VL. This makes it harder to compare performance of different VLMs on vision-language tasks.

**Describe the solution you'd like**

According to the [technical report](https://arxiv.org/pdf/2511.21631), Qwen3-VL adopts a three-module architecture comprising a vision encoder, an MLP-based vision–language merger, and a large language model (LLM). The vision encoder is SigLIP2, which we already have as WIP in #511. 

It's also worth to mention:

* Interleaved MRoPE, which we don't seem to support yet
* DeepStack for the vision-language merger, which is a bit more complicated than what we gave for Gemma 3
* Video timestamp

**Additional context**

A couple of design questions: 

1. Should we wait for #511 to be merged or start the work on Qwen3-VL in parallel? Calling @abheesht17 for an opinion. 
2. Should we extend (text-only) Qwen3 or create a totally new model? I'm not sure it will be easy to integrate DeepStack without changing the way of the text-only version. 

**Checklist**

- [x] I have searched the existing issues for similar feature requests.
- [x] This is not a support question (please use the "bug template" for that).

### PR DESCRIPTION
**Pull request by ridcl**
This PR implements Qwen3-VL and (partially) resolves #1063. 

What's included: 

* Vision and text model
* Parameter loading from Huggingface endpoint
* Model config for Qwen3-VL 4B

What's not included yet: 

* Sampling. Current sampler in Tunix doesn't support vision input. Moreover, Qwen3-VL has notion of `grid_thw` (spatial information about image and video inputs) that requires additional handling. Thus sampling is out of scope for now.
* Sharding for the vision model. 
* Mapping to SGLang and vLLM. The regular Qwen3 model contains it, but I haven't found any docs on the topic.

**Correctness check**

Since we don't have sampling yet, I checked layerwise matching with transformers. The script is [here](https://gist.github.com/ridcl/a715fd92ddd941655bb0c005daa85ee8). The output: 

```
Embedding max diff: 0.629883
Vision tokens  max=0.6299  mean=0.0270
Deepstack[0]   max=0.0469  mean=0.0027
Deepstack[1]   max=0.7568  mean=0.0083
Deepstack[2]   max=0.4941  mean=0.0127
tf_attention_mask is None: False  (will also run with explicit JAX-derived mask to isolate masking)
Layer  0  max=0.9570  mean=0.0281  worst=(seq=93, dim=731)  is_vision=True
Layer  1  max=2.1875  mean=0.0311  worst=(seq=97, dim=0)  is_vision=True
Layer  2  max=2.0000  mean=0.0344  worst=(seq=98, dim=0)  is_vision=True
Layer  3  max=2.3750  mean=0.0346  worst=(seq=98, dim=0)  is_vision=True
Layer  4  max=3.2500  mean=0.0343  worst=(seq=98, dim=0)  is_vision=True
Layer  5  max=4.0000  mean=0.0340  worst=(seq=97, dim=0)  is_vision=True
Layer  6  max=64.0000  mean=0.0341  worst=(seq=1, dim=4)  is_vision=False
Layer  7  max=64.0000  mean=0.0340  worst=(seq=1, dim=4)  is_vision=False
Layer  8  max=64.0000  mean=0.0355  worst=(seq=1, dim=4)  is_vision=False
Layer  9  max=64.0000  mean=0.0344  worst=(seq=1, dim=4)  is_vision=False
Layer 10  max=64.0000  mean=0.0353  worst=(seq=1, dim=4)  is_vision=False
Layer 11  max=64.0000  mean=0.0359  worst=(seq=1, dim=4)  is_vision=False
Layer 12  max=64.0000  mean=0.0385  worst=(seq=1, dim=4)  is_vision=False
Layer 13  max=64.0000  mean=0.0401  worst=(seq=1, dim=4)  is_vision=False
Layer 14  max=64.0000  mean=0.0416  worst=(seq=1, dim=4)  is_vision=False
Layer 15  max=64.0000  mean=0.0435  worst=(seq=1, dim=4)  is_vision=False
Layer 16  max=64.0000  mean=0.0450  worst=(seq=1, dim=4)  is_vision=False
Layer 17  max=64.0000  mean=0.0477  worst=(seq=1, dim=4)  is_vision=False
Layer 18  max=64.0000  mean=0.0520  worst=(seq=1, dim=4)  is_vision=False
Layer 19  max=64.0000  mean=0.0580  worst=(seq=1, dim=4)  is_vision=False
Layer 20  max=64.0000  mean=0.0642  worst=(seq=1, dim=4)  is_vision=False
Layer 21  max=64.0000  mean=0.0700  worst=(seq=1, dim=4)  is_vision=False
Layer 22  max=64.0000  mean=0.0832  worst=(seq=1, dim=4)  is_vision=False
Layer 23  max=64.0000  mean=0.1035  worst=(seq=1, dim=4)  is_vision=False
Layer 24  max=64.0000  mean=0.1362  worst=(seq=1, dim=4)  is_vision=False
Layer 25  max=64.0000  mean=0.1587  worst=(seq=1, dim=4)  is_vision=False
Layer 26  max=64.0000  mean=0.1861  worst=(seq=1, dim=4)  is_vision=False
Layer 27  max=64.0000  mean=0.2099  worst=(seq=1, dim=4)  is_vision=False
Layer 28  max=64.0000  mean=0.2432  worst=(seq=1, dim=4)  is_vision=False
Layer 29  max=64.0000  mean=0.2894  worst=(seq=1, dim=4)  is_vision=False
Layer 30  max=64.0000  mean=0.3507  worst=(seq=1, dim=4)  is_vision=False
Layer 31  max=64.0000  mean=0.4176  worst=(seq=1, dim=4)  is_vision=False
Layer 32  max=64.0000  mean=0.4967  worst=(seq=1, dim=4)  is_vision=False
Layer 33  max=64.0000  mean=0.5858  worst=(seq=1, dim=4)  is_vision=False
Layer 34  max=84.5000  mean=0.6787  worst=(seq=98, dim=4)  is_vision=True
Layer 35  max=64.0000  mean=0.9139  worst=(seq=1, dim=4)  is_vision=False
Final norm max diff: 43.500000
Logits  max=492.4688  mean=7.2643
JAX top-5: [1986, 334, 95456, 32, 108893]
 PT top-5: [1986, 32, 785, 95456, 8420]
Top-1 match: True
```
The discrepancy between activations comes from the difference in XLA vs. cuDNN implementation of matmul (discrepancy in the last digit of bfloat16) and is amplified by MLP layers. Despite seemingly large difference in the output logits, I check that the top-1 token matches between JAX and PyTorch implementations. 

**Reference**

Vision encoder is based on: 
* [sdan/Qwen3-VL-JAX](https://github.com/sdan/Qwen3-VL-JAX) - for the base model.
* [erfanzar/EasyDeL](https://github.com/erfanzar/EasyDeL/blob/main/easydel/modules/qwen3_vl/modeling_qwen3_vl.py)- for mask and some other utils.
* [huggingface/transformers](https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen3_vl/modeling_qwen3_vl.py) - vision-text integration. 

Text decoder is mostly a copy of `tunix/models/qwen3` with RoPE -> mRoPE and deepstack integration. 

**Colab Notebook**

Not used to Colab notebooks, but here's a gist showing the usage with image features:
https://gist.github.com/ridcl/9adb25ecf5a843c3cfae1a9285cf4473

**Checklist**

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and all unit tests pass.
- [x] I have added all appropriate doc-strings/documentation.
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).
- [x] I have followed [Contribution Guidelines](https://github.com/google/tunix/blob/main/docs/contributing.md).

### CODE REVIEW

**Comment by ridcl** [comment]
Could one of the maintainers approve the pipeline please? Since I developed (and tested) the new feature on GPU and the other tests here clearly require TPU, I'd like to make sure that they are still passing in CI/CD.

**Comment by ridcl** [comment]
I've update the PR quite a bit. New files: 

### `consistency_test.py`

Formalized version of the ad-hoc script that I used previously. Requires several PyTorch dependencies to run. Results on 2 x RTX3090:

```
$ PYTHONPATH=. python tunix/models/qwen3vl/consistency_test.py --image_url https://fastly.picsum.photos/id/541/300/200.jpg?hmac=EU9KBKReX22D8zAU9GY1iRAuNDwf5pJa3hyZA2eHiDQ
...
Vision tokens  max=1.0215  mean=0.0365
Deepstack[0]   max=0.0469  mean=0.0033
Deepstack[1]   max=0.4961  mean=0.0095
Deepstack[2]   max=1.0000  mean=0.0171
Embedding max diff (post-injection): 1.021484

Layer    max diff   mean diff  worst_seq  worst_dim  is_vision
-----------------------------------------------------------------
    0      1.1250    0.031615         43          0       True
    1      3.1250    0.032934         43          0       True
    2      3.5625    0.033382         43          0       True
    3      3.9375    0.033657         43          0       True
    4      4.5000    0.033958         43          0       True
    5      4.3750    0.033609         43          0       True
    6     16.0000    0.033477          1        396      False
    7     16.0000    0.033544          1        396      False
    8     16.0000    0.034278          1        396      False
    9     16.0000    0.034796          1        396      False
   10     16.0000    0.036017          1        396      False
   11     16.0000    0.039787          1        396      False
   12     16.0000    0.041447          1        396      False
   13     16.0000    0.043618          1        396      False
   14     16.0000    0.044483          1        396      False
   15     16.0000    0.047405          1        396      False
   16     16.0000    0.049590          1        396      False
   17     16.0000    0.056943          1        396      False
   18     16.0000    0.063522          1        396      False
   19     16.0000    0.070109          1        396      False
   20     16.0000    0.074904          1        396      False
   21     16.0000    0.079529          1        396      False
   22     16.0000    0.088248          1        396      False
   23     16.0000    0.106456          1        396      False
   24     16.0000    0.124820          1        396      False
   25     16.0000    0.139814          1        396      False
   26     18.7500    0.164371         43          4       True
   27     22.5000    0.189085         43          4       True
   28     27.5000    0.230367         43          4       True
   29     30.7500    0.276541         43          4       True
   30     27.0000    0.342424         43          4       True
   31     26.0000    0.421054         43          4       True
   32     23.0000    0.507756         43          4       True
   33     24.5000    0.595446         10          0       True
   34     44.0000    0.679490         10          4       True
   35     53.0000    0.882008         20          4       True

Final norm max diff: 16.937500
Logits  max=1.5000  mean=0.2163
JAX top-5 tokens: [785, 4792, 40, 1986, 7771]
 PT top-5 tokens: [785, 4792, 40, 1986, 2073]
Top-1 match: True
```

### `utils.py`

Several helpers to make work with vision-enabled Qwen smoother. Includes:

* `load_processor()` - workaround for `AutoProcessor.from_pretrained()` when torchvision is not installed. A bit weird, but I couldn't come up with a better design yet.
* A few utils for batch encoding. Think of them as `AutoProcessor.__call__()`, but more JAX-friendly. Comments and docstrings give some details, but the root problem was that the original implementation in PyTorch calculates positional embeddings (MRoPE) from `grid_thw` (metadata about images) in a very dynamic way incompatible with JAX's JIT. So instead I move calculation of all static parameters to the data preparation step and wrap into `VisionGridData` dataclass. Where in Huggingface one would write: 

```python
processor = AutoProcessor.from_pretrained(...)
inputs = processor(text=..., images=...)
```
In this implementation, one should use:

```python
processor = load_processor(...)   # Same processor, but in a torch-free environment
inputs = encode_batch(processor, texts=..., images=...)
```

### `sampler.py`

The vision data makes Qwen3-VL incompatible with the main sampler, and because of the peculiarities above it's not obvious how to adapt the sampler in future-proof way. `sampler.py` provides a Qwen-only, image-enabled alternative. Again, a bit weird, but at least we can now ensure that the model doesn't generate garbage. 

### `train_example.py`

An example script that combines all of the above and makes sure it's compatible with the rest of the ecosystem. For now, I treat it more as a test of the implementation, but once the API stabilizes we probably want to `examples/`. 

Tagging a few people at random to bring attention to this PR: 
@abheesht17 @tianshub @jiangyangmu

**Comment by ridcl** [comment]
As a side note, [I also tried to build Qwen3.5](https://github.com/ridcl/tunix/blob/qwen3_5/tunix/models/qwen3_5/model.py) in a fully automatic way (it turns out LLMs are pretty good in PyTorch -> JAX translation). The implementation is technically correct, but inefficient. The original version interleaves dot-product and flash linear attention. Flash linear attention is implemented as a dedicated kernel in Triton (forward pass only). I checked if it's possible to rewrite it to Pallas, but I don't know either of these technologies well enough to really make educated judgement.
